fix(portal): Prevent races during relay selection by only using the ones connected for more than 5 seconds ago

AndrewDryga opened the pull request:
Closes #6099
Should push #6109 to not being needed short term.

AndrewDryga committed:
Prevent races during relay selection by only using the ones connected more than 5 seconds ago

jamilbk reviewed (approved):
LGTM, just want to note that this and the consistency fix are included in this one PR right?

One thought that came to mind here was if we roll over Relays too quickly, but I guess the region deployment strategy should prevent that, i.e. one instance in each region is rolled.

AndrewDryga committed:
Add timeouts for relays to become active to integration and perf tests

AndrewDryga commented on elixir/apps/api/lib/api/client/channel.ex:
```diff
@@ -327,7 +327,9 @@ defmodule API.Client.Channel do
-        {:ok, relays} = select_relays(socket)
+        :ok = Relays.unsubscribe_from_relay_presence(relay_id)
+
+        {:ok, relays} = select_relays(socket, [relay_id])
```
Here is the additional fix for relay being present in both lists.

thomaseizinger commented on elixir/apps/api/test/api/client/channel_test.exs:
```diff
@@ -333,18 +350,18 @@ defmodule API.Client.ChannelTest do
+      assert relay_view1.id == relay2.id
+      assert relay_view2.id == relay2.id
+      assert relay1_id == relay1.id
     end
```
The property that we are interested in, and which currently doesn't hold is:

```suggestion
        assert relay_view1.id != relay1_id
        assert relay_view2.id != relay1_id
    end
```

It might be worth adding 1 more test that represents staging / prod more closely where we have ~10 relays, disconnect them all one by one, connect new ones, capture all relays presence messages and assert that none of the disconnected ones ever get returned as connected.

That test doesn't need to capture which specific ones of the newly connected ones do get returned so you don't need to simulate the selection logic. We use property-based testing connlib extensively and it helped us find a lot of these subtle bugs :)

AndrewDryga replied:
Isn't
```
assert relay_view1.id == relay2.id
```
is the same as
```
assert relay_view1.id != relay1_id
```
because we know that `relay2.id != relay1_id`?

> It might be worth adding 1 more test that represents staging / prod more closely where we have ~10 relays, disconnect them all one by one, connect new ones, capture all relays presence messages and assert that none of the disconnected ones ever get returned as connected.

The suggested test is the same as restarting this one multiple times, before we were testing on just one relay, I've added 2nd to capture that (doesn't matter if it's 10 of them or 2). So if there is a bug our tests will find that sooner than later anyways.

I will add a test that makes sure that relay won't be ever returned even if it was single one online though.

thomaseizinger replied:
> Isn't
> 
> ```
> assert relay_view1.id == relay2.id
> ```
> 
> is the same as
> 
> ```
> assert relay_view1.id != relay1_id
> ```
> 
> because we know that `relay2.id != relay1_id`?

In this test setup, yes. If you'd use more inputs, modelling which ones is actually returned gets more difficult and obscures the test. In connlib, we don't care which one we get but the property of not receiving a disconnected one is really important. What I am suggesting is IMO simply a more intentional assertion of what we want. It is your codebase, so up to you :)

thomaseizinger reviewed (approved):
Thanks for the quick turnaround! Approving to unblock. I can't fully confirm that is fixes the bug because I can't read elixir well enough. A property-based test could help though!

AndrewDryga committed:
Increase timeouts for relay restart tests

AndrewDryga committed:
Add a test to make sure relay is never returned in both elements of the presence

thomaseizinger commented on elixir/apps/domain/lib/domain/relays.ex:
```diff
@@ -244,12 +246,13 @@ defmodule Domain.Relays do
       Relay.Query.not_deleted()
       |> Relay.Query.by_ids(connected_relay_ids)
       |> Relay.Query.global_or_by_account_id(account_id)
+      |> Relay.Query.by_last_seen_at_greater_than(5, "second", :ago)
```
Do we really need this if we filter the currently disconnecting one? Can you explain where the race condition actually happens?

AndrewDryga replied:
@thomaseizinger I added this as a preventive measure from too quickly changing the online state, it's not required to fix the race. Will think for a bit if we should go with it or not.

AndrewDryga committed:
Remove relays from connection request flow

AndrewDryga committed:
Remove delay for connected relays

This adds a pain with bash-based CI jobs

AndrewDryga committed:
Make sure clients will get at least one relay

Now we subscribe clients and gateways to global and account relays presence (once) whenever we return empty relays list

AndrewDryga committed:
Remove sleeps
